server/src/main/java/org/opensearch/action/search/SearchTask.java

```diff
@@ -48,7 +49,12 @@
 
     public SearchTask(long id, String type, String action, Supplier<String> descriptionSupplier,
                       TaskId parentTaskId, Map<String, String> headers) {
-        super(id, type, action, null, parentTaskId, headers);
+        this(id, type, action, descriptionSupplier, parentTaskId, headers, false, null);
+    }
+
+    public SearchTask(long id, String type, String action, Supplier<String> descriptionSupplier,
+        TaskId parentTaskId, Map<String, String> headers, boolean cancelOnTimeout, TimeValue cancelTimeout) {
+        super(id, type, action, null, parentTaskId, headers, cancelOnTimeout, cancelTimeout);
```

Comment:
> `cancelOnTimeout, cancelTimeout` these are confusing to me, can you explain?
> is the `cancelOnTimeout` a boolean whether or not to use the new param and the `cancelTimeout` the interval we will take into account if `cancelOnTimeout` is `true` ?

Yes, your understanding in right. May be some context will help. For each request, a task object is created and registered in the TaskManager at transport layer. The reason I did it in this way is because `SearchTask` is used by multiple request type:

1. [SearchRequest](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/action/search/SearchRequest.java#L674)
2. [SearchScrollRequest](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/action/search/SearchScrollRequest.java#L134)
3. [AsyncSearchRequest](https://github.com/opensearch-project/asynchronous-search/blob/bd31e80adf6d52c1b4662d0d2cc9b30d8ae14309/src/main/java/org/opensearch/search/asynchronous/task/AsynchronousSearchTask.java#L44)

We don't want to support cancellation after timeInterval for all such requests, since 2 and 3 are expected to be long running requests and have their own cancellation mechanism at request level. However, since they all are using same `SearchTask` class I need a mechanism to differentiate it being created by specific request type which supports timeout based cancellation (hence the flag). I thought about other approaches as well: 

1. Create a new `CancellableTask` type like `CancellationOnTimeoutTask` and let `SearchTask` extends that. But that would mean all the `SearchTask` supports cancellation on timeout, which we don't want.
2. Create a new task type like `SearchCancellationOnTimeoutTask` which would extends `CancellableTask` type. Now for case 1 only, search request will create the task of this new type. All other cases, will keep using existing SearchTask type. However, in search code path I am seeing multiple places where the Task is type-casted to `SearchTask` which will break with this approach.
3. Create a new task type like `SearchCancellationOnTimeoutTask` that extends `SearchTask` type. And for case 1 only, SearchRequest will create the task of this new type. Casting issue in 2 will be solved, however, now if we want to support cancellation on timeout for other request then a new task type for each such request needs to be created. With current approach since `CancellableTask` itself has these 2 params the existing task type just needs to support the new constructor.

> It seems that the boolean may be redundant - and you could relate to `cancelTimeout`'s null as `cancelOnTimeout` = `false`. Does that make sense?
> 
> I think the naming could be more descriptive.
The flag is needed here, since `cancelTimeout == null` doesn't dictate if cancel task needs to be scheduled or not. This task object just stores the request level `cancel_after_time_interval` parameter value if provided. If request level parameter is absent, then backend falls back to cluster level setting value. So if request level parameter is absent we still want to cancel the request using the cluster level setting value. However, as explained above, we also need a mechanism to decide whether cancellation should be applied for a search task or not. This is because same `SearchTask` is used by multiple request type and we only want to support the standard `SearchRequest` type.

To summarize:

`SearchTask` created by standard `SearchRequest` should support cancellation after a time interval with optional request level parameter.
`SearchTask` created by `AsynchronousSearch` should not support cancellation after a time interval.
`SearchTask` created by `SearchScrollRequest` should not support cancellation after a time interval.

Reply:
Got it. So the boolean on the task level is definitely a must. Maybe change the naming to be more descriptive. Using an `is` prefix maybe?

Reply:
Did you consider implementing a `getTimeout()` on each `Request`, and invoking that method instead of passing a `boolean` or timeout values around into a task? This way you can express exactly what you are spelling in your summary? Would checking cancelability in `TransportSearchAction#doExecute` from the request available as input work?

Reply:
> Got it. So the boolean on the task level is definitely a must. Maybe change the naming of the boolean to be more descriptive. Using an `is` prefix maybe?

how about renaming it as `supportCancelOnTimeout` ?

Reply:
> Did you consider implementing a `getTimeout()` and/or `isCancellable` on each `Request` type, and invoking that method instead of passing a `boolean` or timeout values around into a task? This way you can express exactly what you are spelling in your summary? Would checking cancelability in `TransportSearchAction#doExecute` from the request available as input work?

That is exactly what is done in `Task` and that seems to be the right place as internally its the task which is getting cancelled. With request level approach:

1. the boolean will be moved to the request class, this is because `AsyncSearch` internally uses the same `SearchRequest` to trigger actual search. O
2. `Task` is not a serializable object, it is local whereas requests are serializable and we may need to support that for flag as well.
3. All the request doesn't have a common interface to fetch the boolean and cancel timeout value, which is provided by `CancellableTask` interface that is common across all the requests.

Reply:
Will leave it to you to decide, but I would give this approach a try as it might make things cleaner. I think (3) is actually a miss, because we're saying requests can be cancelable or not, while a task is just a mechanical executor of a request.

Reply:
Request and task have 1:1 mapping. Each request owns/create a task for itself and it is the task which controls whether the request is cancellable or not. There is no concept of cancelling a request other than cancelling a task. Any cancellation related information is eventually owned by a Task, which is what this new parameter is also doing.

I found another way to get away with the boolean. The default value of `cancellationTimeout` in `CancellableTask` can be set to -1 or `NO_TIMEOUT`. Now requests which uses `SearchTask` can explicitly set the time interval to a valid value while creating the task object if they want this mechanism to kick in. This way, any consumers like `AsyncSearch` and `SearchScrollRequest` which consumes `SearchTask` will not have to change since default will still be -1 for those. For standard `SearchRequest` we will create task with the request level parameter (if present) or null (which is default) otherwise.